nixos/modules/services/security/opensnitch.nix

```diff
@@ -0,0 +1,174 @@
+        description = ''
+          List of destination hosts for which to create default (regexp) allow rules,
+          regardless of other connection properties.
+        '';
+        default = [];
+        example = ''[ { host = "*.nixos.org" } { ip = "127.0.0.1" } ]'';
+      };
+
+      extraRules = mkOption {
+        type = types.attrsOf types.attrs;
+        description = ''
+          Set of JSON attribute sets describing default opensnitch rules that are written to
+          /etc/opensnitch/rules.
+        '';
+        default = [];
+        example = ''
+          {
+            kerberos_rule = {
+              action = "allow";
+              operator = {
+                type = "list";
+                list = [
+                  {
+                    type = "simple";
+                    operand = "dest.host";
+                    data = "kerberos.example.domain";
+                  }
+                  {
+                    type = "simple";
+                    operand = "dest.port";
+                    data = "88";
+                  }
+                ];
+              };
+            };
+          }
+        '';
+      };
+    };
+  };
+
+  config =
+    let
+      uiConfig = writeText "ui-config.json" (builtins.toJSON cfg.uiConfig);
+    in
+    lib.mkIf cfg.enable {
+
+      environment.etc = listToAttrs (
+        (map (makePackageRuleFile "opensnitchd/rules") cfg.whitelistPackages)
+        ++ concatMap (x: mapAttrsToList (n: v: makeHostRuleFile "opensnitchd/rules" n v) x) cfg.whitelistHosts
+        ++ mapAttrsToList (makeAlwaysRuleFile "opensnitchd/rules") cfg.extraRules
+      );
+
+      environment.systemPackages = if cfg.startUserService then [] else [ opensnitch-ui ];
+
+      systemd.services.opensnitchd = {
+        description = "opensnitch firewall daemon";
+        after = [ "network.target" ];
+        wants = [ "network.target" ];
+        wantedBy = [ "multi-user.target" ];
+        preStart = ''
+          mkdir -p /etc/opensnitchd/rules
+          chown root:root /etc/opensnitchd
+          chmod 700 /etc/opensnitchd
+          # mkdir -p /run/opensnitch
+        '';
+        path = [ iptables ];
+        # For some reason there is a problem with specifying /run for the socket file...
+        # script = "${lib.getBin opensnitch.daemon}/bin/opensnitchd -log-file /var/log/opensnitchd.log -rules-path /etc/opensnitchd/rules -ui-socket unix:///run/opensnitch/osui.sock -debug";
+        script = "${lib.getBin opensnitchd}/bin/opensnitchd -log-file /var/log/opensnitchd.log -rules-path /etc/opensnitchd/rules -ui-socket unix:///tmp/osui.sock";
+        serviceConfig = {
+          Type = "simple";
+          Restart = "always";
+          RestartSec = "30";
+        };
+      };
+
+      systemd.user.services.opensnitch-ui = {
+        description = "opensnitch firewall UI process";
+        wantedBy = [ "graphical-session.target" ];
+        partOf = [ "graphical-session.target" ];
```

Comment:
Oh?

Reply:
Yeah this is the graphical client application which runs in the background of the user session and displays the popup windows. Maybe we should start by just packaging the opensnitch daemon only?

Reply:
I just meant that if this is `partOf` then the graphical target is marked as failed if this one package fails. Usually we just use `wantedBy` instead for software that isn't part of the unit.